### PR TITLE
Improve yamux; add test-plans interop app

### PR DESCRIPTION
**Pull request by flcl42**
Added in the PR:
- Protocols:
  - Add half closing, basic window support to yamux
  - Add partial support of multiplexer fast selection in noise
- Interop
  - Add add test-plans interop app to samples
  - Fully fix TCP; improve QUIC
- Other
  - Redesign reading and writing: add result codes, end of writing
  - Fix optional dependency

### CODE REVIEW

File: src/libp2p/Libp2p.Core.TestsBase/DebugLoggerFactory.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
```
Comment (rubo) [inline src/libp2p/Libp2p.Core.TestsBase/DebugLoggerFactory.cs:1]: Minor: For new file headers, please use 2024.

File: src/libp2p/Libp2p.Core.TestsBase/TestChannel.cs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 // SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
 // SPDX-License-Identifier: MIT
 
+using Newtonsoft.Json.Linq;
```
Comment (rubo) [inline src/libp2p/Libp2p.Core.TestsBase/TestChannel.cs:4]: Why Newtonsoft? Can't we use System.Text.Json?

File: src/libp2p/Libp2p.Core/IWriter.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@ namespace Nethermind.Libp2p.Core;
 
 public interface IWriter
 {
-    ValueTask WriteLineAsync(string str, bool prependedWithSize = true)
+    ValueTask<IOResult> WriteLineAsync(string str, bool prependedWithSize = true)
```
Comment (rubo) [inline src/libp2p/Libp2p.Core/IWriter.cs:12]: What's the reason for `ValueTask` here?

Comment (flcl42) [reply src/libp2p/Libp2p.Core/IWriter.cs:12]: It's instantly completed most of the times, which is more efficient to handle with ValueTask. There are mid term plans to run and share benchmarks.

**Review by rubo — APPROVED** [review]

**Pull request merged**